martin/src/cog/image.rs

```diff
@@ -0,0 +1,228 @@
+fn rgb_to_png(
+    vec: Vec<u8>,
+    (tile_width, tile_height): (u32, u32),
+    (data_width, data_height): (u32, u32),
+    chunk_components_count: u32,
+    nodata: Option<u8>,
+    path: &Path,
+) -> Result<Vec<u8>, CogError> {
+    let is_padded = data_width != tile_width || data_height != tile_height;
+    let need_add_alpha = chunk_components_count != 4;
+
+    let pixels = if nodata.is_some() || need_add_alpha || is_padded {
+        let mut result_vec = vec![0; (tile_width * tile_height * 4) as usize];
+        for row in 0..data_height {
+            'outer: for col in 0..data_width {
+                let idx_chunk =
+                    row * data_width * chunk_components_count + col * chunk_components_count;
+                let idx_result = row * tile_width * 4 + col * 4;
+                for component_idx in 0..chunk_components_count {
+                    if nodata.eq(&Some(vec[(idx_chunk + component_idx) as usize])) {
+                        //This pixel is nodata, just make it transparent and skip it then
+                        let alpha_idx = (idx_result + 3) as usize;
+                        result_vec[alpha_idx] = 0;
+                        continue 'outer;
+                    }
+                    result_vec[(idx_result + component_idx) as usize] =
+                        vec[(idx_chunk + component_idx) as usize];
+                }
+                if need_add_alpha {
+                    let alpha_idx = (idx_result + 3) as usize;
+                    result_vec[alpha_idx] = 255;
+                }
+            }
+        }
+        result_vec
+    } else {
+        vec
+    };
+    let mut result_file_buffer = Vec::new();
+    {
+        let mut encoder = png::Encoder::new(
+            BufWriter::new(&mut result_file_buffer),
+            tile_width,
+            tile_height,
+        );
+        encoder.set_color(png::ColorType::Rgba);
+        encoder.set_depth(png::BitDepth::Eight);
+        let mut writer = encoder
+            .write_header()
+            .map_err(|e| CogError::WritePngHeaderFailed(path.to_path_buf(), e))?;
+        writer
+            .write_image_data(&pixels)
+            .map_err(|e| CogError::WriteToPngFailed(path.to_path_buf(), e))?;
+    }
+    Ok(result_file_buffer)
+}
+
+#[cfg(test)]
+mod tests {
+    use std::path::PathBuf;
+
+    use martin_tile_utils::TileCoord;
+    use rstest::rstest;
+
+    use crate::cog::image::Image;
+
+    #[test]
+    fn can_calc_tile_idx() {
+        let image = Image {
+            ifd: 0,
+            across: 3,
+            down: 3,
+            nodata: None,
+        };
+        assert_eq!(Some(0), image.get_tile_idx(TileCoord { z: 0, x: 0, y: 0 }));
+        assert_eq!(Some(8), image.get_tile_idx(TileCoord { z: 0, x: 2, y: 2 }));
+        assert_eq!(None, image.get_tile_idx(TileCoord { z: 0, x: 3, y: 0 }));
+        assert_eq!(None, image.get_tile_idx(TileCoord { z: 0, x: 1, y: 9 }));
+    }
+    #[rstest]
+    // the right half should be transprent
```

Comment:
spelling here and below

```suggestion
    // the right half should be transparent
```